src/v/cloud_topics/level_zero/gc/level_zero_gc.cc

```diff
@@ -0,0 +1,231 @@
+                }
+            }
+
+        } catch (...) {
+            vlog(
+              cd_log.info,
+              "Level zero GC restarting after error: {}",
+              std::current_exception());
+            backoff = config_.throttle_no_progress;
+        }
+    }
+
+    vlog(cd_log.info, "Level zero GC worker is exiting");
+}
+
+seastar::future<std::expected<size_t, level_zero_gc::collection_error>>
+level_zero_gc::try_to_collect() {
+    const auto candidate_objects = co_await storage_->list_objects(&asrc_);
+    if (!candidate_objects.has_value()) {
+        vlog(
+          cd_log.debug,
+          "Received error listing objects during L0 GC: {}",
+          candidate_objects.error());
+        co_return std::unexpected(collection_error::service_error);
+    }
+
+    const auto maybe_max_gc_epoch
+      = co_await epoch_source_->max_gc_eligible_epoch(&asrc_);
+    if (!maybe_max_gc_epoch.has_value()) {
+        vlog(
+          cd_log.debug,
+          "Received error retrieving GC eligible epoch: {}",
+          maybe_max_gc_epoch.error());
+        co_return std::unexpected(collection_error::service_error);
+    }
+
+    const auto max_gc_epoch = maybe_max_gc_epoch.value();
+    if (!max_gc_epoch.has_value()) {
+        vlog(cd_log.info, "No GC eligible epoch currently exists");
+        co_return std::unexpected(collection_error::no_collectible_epoch);
+    }
+
+    const auto max_gc_birthday = std::chrono::system_clock::now()
+                                 - config_.deletion_grace_period;
+
+    // objects that can be safely deleted
+    std::vector<cloud_storage_clients::client::list_bucket_item>
+      eligible_objects;
+
+    // used to detect unsorted object listings
+    seastar::sstring last_key;
+    std::optional<cluster_epoch> last_epoch;
+
+    for (const auto& object : candidate_objects.value().contents) {
+        const auto object_epoch = object_path_factory::level_zero_path_to_epoch(
+          object.key);
+
+        // validate expected L0 object name format, and extract epoch
+        if (!object_epoch.has_value()) {
+            vlog(
+              cd_log.error,
+              "Unable to parse epoch during L0 GC: {}",
+              object_epoch.error());
+            co_return std::unexpected(collection_error::invalid_object_name);
+        }
+
+        // detect non-lexiographic ordering. this may indicate that GC will not
+        // operate efficiently with the underlying storage system. see the class
+        // comment for more details about what this means in practice.
+        if (!last_epoch.has_value()) {
+            last_key = object.key;
+            last_epoch = object_epoch.value();
+        }
+
+        if (object_epoch.value() < last_epoch) {
+            constexpr std::chrono::minutes rate_limit{1};
+            static seastar::logger::rate_limit rate(rate_limit);
+            vloglr(
+              cd_log,
+              seastar::log_level::warn,
```

Comment:
Should this be an error so we can detect it in CI?

I guess it doesn't necessarily hurt correctness, but I can imagine in the worst case of non-lexicographic ordering, this loop could get stuck iterating over new objects, and never make progress onto the oldest objects